Added metrics from the apache scoreboard

daanleenders opened the pull request:
### What does this PR do?

The PR adds metrics that can be derived from the Apache scoreboard. Mainly the maximum of workers available and a count of workers for each state a worker can be in. This is an expansion on the existing `idle_workers` and `busy_workers` metrics that are currently available. 

### Motivation
Currently datadog reports the busy and idle workers but those metrics are most usefull in the context of the maximum workers that can be started on a server. For example, you could have 100 Busy workers but if you have a heavy webserver that can easily handle 500 workers you are fine. But when you run with 100 max workers your apache is full and putting requests in a queue or maybe even dropping requests. The idle workers will then be 0 and you can set-up an alert on that but there are cases where the idle_workers will be 0 but the max workers aren't reached. A monitor of busy workers as a percentage of the max workers would be a more robuste metric and that was the initial motivation to expand the check. 

When I was investigating the way to find this metric on the `mod_status` page it became clear that I would need to check the scoreboard for this. The scoreboard gives us a detailed look at al the (possible) workers and I thought it wouldn't be that difficult to add all the different states in the scoreboard as separate metrics. 

This is very usefull because it allows you to write more precise monitor messages. If you know you have too many busy workers you will always instruct someone to go look at the scoreboard to see if your workers are all busy sending replies or maybe all waiting with a keepalive. These two have very different causes and need different steps to resolve. It would be awesome if the monitor would already detect this and have the title and message reflect this. 

And the different states could also make for a awesome timeseries where the different states would be displayed in a stacked area. This would allow you to quickly spot i which state the bulk of your workers are and see this over time. The latter is lacking from the apache status page since it will only show you the scoreboard at the moment you load/refresh the page, but you won't know what it looked like minutes/seconds ago. 

### Additional Notes

Example of a scoreboard. Each character represents a (potential) worker. 
```
WWWWWWWWW_WW_WWW_.WW..WWWW_W...WWWW.._WWWW...W.WW.W.......G.....
................................................................

Scoreboard Key:
"_" Waiting for Connection, "S" Starting up, "R" Reading Request,
"W" Sending Reply, "K" Keepalive (read), "D" DNS Lookup,
"C" Closing connection, "L" Logging, "G" Gracefully finishing,
"I" Idle cleanup of worker, "." Open slot with no current process
```

I checked the source code of `mod_status` to see when a status counts in the existing `idle_workers` and `busy_workers` metrics and you can see that there are statusses where apache won't be able to utilize/create a worker but it won't be counted as `busy` or `ready` yet. I think it would also help to use this to inform people in the documentation how these new metrics compare to the existing ones. 

Scoreboard key | Explanation on status page | ?notable text | Source code constant | Counted as idle (ready) | Counted as busy (busy)
-- | -- | -- | -- | -- | --
_ | Waiting for Connection | Ready | SERVER_READY | 1 | 0
S | Starting up | Starting | SERVER_STARTING | 0 | 0
R | Reading request | Read | SERVER_BUSY_READ | 0 | 1
W | Sending Reply | Write | SERVER_BUSY_WRITE | 0 | 1
K | Keepalive (read) | Keepalive | SERVER_BUSY_KEEPALIVE | 0 | 1
D | DNS Lookup | DNS lookup | SERVER_BUSY_DNS | 0 | 1
C | Closing connection | Closing | SERVER_CLOSING | 0 | 1
L | Logging | Logging | SERVER_BUSY_LOG | 0 | 1
G | Gracefully finishing | Graceful | SERVER_GRACEFUL | 1 _(when async)_ | 1 _(when not async)_
I | Idle cleanup of worker | Dying | SERVER_IDLE_KILL | 0 | 0
. | Open slot with no current process | Dead | SERVER_DEAD | 0 | 0
" " | not mentioned | omitted | SERVER_DISABLED | 0 | 0

_source: http://svn.apache.org/viewvc/httpd/httpd/trunk/modules/generators/mod_status.c?view=markup#l327 , http://svn.apache.org/viewvc/httpd/httpd/trunk/modules/generators/mod_status.c?view=markup#l777 & http://svn.apache.org/viewvc/httpd/httpd/trunk/modules/generators/mod_status.c?view=markup#l1002_

### Review checklist (to be filled by reviewers)

- [ ] Feature or bugfix MUST have appropriate tests (unit, integration, e2e)
- [ ] PR title must be written as a CHANGELOG entry [(see why)](https://github.com/DataDog/integrations-core/blob/master/CONTRIBUTING.md#pull-request-title)
- [ ] Files changes must correspond to the primary purpose of the PR as described in the title (small unrelated changes should have their own PR)
- [ ] PR must have `changelog/` and `integration/` labels attached

FlorianVeaux commented on apache/datadog_checks/apache/apache.py:
```diff
@@ -29,6 +30,21 @@ class Apache(AgentCheck):
 
     RATES = {'Total kBytes': 'apache.net.bytes_per_s', 'Total Accesses': 'apache.net.request_per_s'}
 
+    SCOREBOARD_KEYS = {
```
I see two possible approaches for reporting this data.
1. The one that you implemented where each "worker status" gets its own metric. The benefits are clear, though it's a bit hard if you want to get all metrics in an area graph.
2. The second approach would be to report a single metric (let's say `apache.scoreboard.workers_by_status`) that is tagged by its `worker_status`. The benefit of this approach is that it'll be much easier to make complicated queries like 
    a. Show an area graph with all the workers status.
    b. Show the total count of workers that are not disabled.
The drawback is that if you graph the plain metric `apache.scoreboard.workers_by_status` with the default `avg` space-aggregator, you'd get the average number of worker per status which doesn't make a lot of sense

The first option is perfectly fine in the general case but as you rightfully the mentioned the benefit of having a stacked-area graph, I believe the second option would work best.

daanleenders replied:
It was indeed an implementation detail I wasn't sure about. The reason I implemented 1 instead of 2 was because of other metrics I am familiar with that (to me at least) are similar and are implemented with each state/type as a separate metric. For example the CPU usage is split in 6 different metrics (`system.cpu.idle`, `system.cpu.user`, etc.). And the mysql queries (`mysql.performance.queries`) only reports the total queries and each query type isn't a tag but a separate metric (`mysql.performance.com_select`, `mysql.performance.com_insert`, etc.). 

The only example of metrics that I use and could think of while writing this are the haproxy metrics that tag each metric with a different `haproxy_service` to indicate which frontend/backend in haproxy the metric belongs to. In apache this would translate to different vhosts but on the `?auto` version of the mod status page you aren't able to know which vhost a worker is busy with. 

Thinking about the examples I knew of other integrations doing something similar it seemed that using different metrics for each state instead of tagging the states would be preferred. 

> b. Show the total count of workers that are not disabled.

Yeah excluding one or some states would be a lot easier with tagging, than it would be by combining metrics. But I feel like the discoverability of the different states is a lot harder, and as you stated the data you would get by not using the tags would be garbage by default. So when you were just trying each metric that the Apache integration provides by adding them to a timeseries widget in a dashboard (what I personally do to find useful metrics) it wouldn't be clear what its use is. 

> though it's a bit hard if you want to get all metrics in an area graph.

I feel like this could be easily fixed by adding an area timeseries with each state visible to the preset Apache overview dashboard. If a user doesn't know about the 'Advanced...' button to create stacked areas from different metrics, they could just copy the timeseries and paste it in their own dashboard. 

> The first option is perfectly fine in the general case but as you rightfully the mentioned the benefit of having a stacked-area graph, I believe the second option would work best.

I believe the best option would be the one that is the most consistent across other integrations. I am more aware of integrations that use the pattern from option 1 for metrics/values like this. But if 2 is used more consistently across integrations or preferred for newer metrics, I would change it to tags instead of separate metrics.

FlorianVeaux replied:
We have exemple of both usage and I agree with you, option 1 is a bit more common as it's easier to make sense of it at the cost of a slight UX issue with stacked area graphs.

I'm perfectly fine with option 1, let's leave it like that 👍

FlorianVeaux commented on apache/datadog_checks/apache/apache.py:
```diff
@@ -81,6 +97,12 @@ def check(self, instance):
             if len(values) == 2:  # match
                 metric, value = values
+
+                # Special case
```
```suggestion
                # Special case: Report the status of the workers
```

FlorianVeaux commented on apache/datadog_checks/apache/apache.py:
```diff
@@ -141,3 +163,19 @@ def _submit_metadata(self, value):
         self.set_metadata('version', version, scheme='parts', final_scheme='semver', part_map=version_parts)
         self.log.debug("found apache version %s", version)
+
+    def _submit_scoreboard(self, value, tags):
+        max_workers = 0
```
```suggestion
    def _submit_scoreboard(self, value, tags):
        """The scoreboard is a long string where each character represents the status of a given worker.
        This method parses that string and emits the corresponding metrics"""
        max_workers = 0
```

FlorianVeaux commented on apache/datadog_checks/apache/apache.py:
```diff
@@ -141,3 +163,19 @@ def _submit_metadata(self, value):
+
+    def _submit_scoreboard(self, value, tags):
+        max_workers = 0
+        metrics = {}
+        for _key, metric in self.SCOREBOARD_KEYS.items():
+            metrics[metric] = 0
+
+        for i in value:
+            max_workers += 1
+            metrics[self.SCOREBOARD_KEYS[i]] += 1
+
+        for _key, metric in self.SCOREBOARD_KEYS.items():
+            self.gauge(metric, metrics[metric], tags=tags)
```
Small implementation nit, we can replace that code by the following one which is easier to understand

```suggestion
        for _key, metric in self.SCOREBOARD_KEYS.items():
            self.gauge(metric, value.count(_key), tags=tags)
        
        self.gauge(self.GAUGES['MaxWorkers'], len(value), tags=tags)
        self.log.debug("Scoreboard: %s", value)
```

Additionally we might want to add a validation check at the beginning of that method to make sure that the `value` is a correctly formatted scoreboard. Something like this

```python
if not all(c in self.SCOREBOARD_KEYS.keys() for c in value):
    self.log.info("Invalid scoreboard value, ignoring the scoreboard metrics")
    return
```

daanleenders replied:
> Additionally we might want to add a validation check at the beginning of that method to make sure that the value is a correctly formatted scoreboard. Something like this

I'm hesitant to put something like this in because it would completely stop reporting any of the new worker metrics when a newer version of Apache would add a new state. Personally I would prefer having the known metrics and only missing a new state when Apache decides to add one, instead of losing all the metrics I have added in this PR. Given that scoreboard is a core mechanism of Apache I can't imagine a day where one of there status boards will report a different kind of value under the name 'scoreboard'. So I would expect that the change of receiving a value that doesn't adhere to the expected format is lower than the change of them adding an extra state to the existing format. 

I don't have a strong opinion on this and would gladly put a check like that (probably exactly that check😉) in but this is why I haven't done it initially.

FlorianVeaux replied:
The issue with Apache adding a new state was that your old code would raise an exception when trying to read from SCOREBOARD_KEYS, adding a pre-validation was a nice nit in that situation to give the user a more meaningful error message than a big stack trace.

By using `count`, we shouldn't raise any exception anymore. The code gracefully handles unknown character by ignoring it. Thus making the validation unneeded.

daanleenders replied:
Oh, off course. That's what you get from building in PHP on your day job 😅 , PHP doesn't care. I thought/hoped that sticking to the basics would prevent me from making those kinds of errors, but apparently it won't.

FlorianVeaux reviewed (changes requested):
Thanks @daanleenders
This is a great idea.

There is one open question with whether or not to send the various worker status as a tag or as distinct metrics. I've left a few details in my comment below and I'd like to hear your thoughts on this.

daanleenders commented:
> Thanks @daanleenders
> This is a great idea.
> 
> There is one open question with whether or not to send the various worker status as a tag or as distinct metrics. I've left a few details in my comment below and I'd like to hear your thoughts on this.

You're welcome. Thank you for looking at it, I personally really mis max_workers and would love to have it available in production :). 

I have responded to the 2 questions in your review with my reasons for the choices I made with my initial proposal. My whole PR should be interpreted like that, an initial proposal. I am not aware of best practices in creating metrics nor that great in writing python.

p.s. I just had some issues with running the existing tests. The `test_connection_failure` test would fail because it could successfully connect with `localhost:1234`. I was still running a instance of [parcel](https://parceljs.org/) which exposes itself by default on `localhost:1234` so it was in fact available. Should I create a separate PR to change the test to point to something that should never resolve? I would propose something like `bad.config.invalid:1234` or just `invalid:1234` since the ".invalid" TLD is [meant for examples/tests like this](https://datatracker.ietf.org/doc/html/rfc2606#section-2)

FlorianVeaux commented:
You seem to have included unrelated changes in your PR.
Besides that the PR looks good to go for me, thanks for discussing the other options your implementation choices look good for me.

I'll let you clean the unrelated commit and we can merge it soon after!

daanleenders commented:
> You seem to have included unrelated changes in your PR.

I see it. Probably did something wrong with rebasing master. I'll probably have time to fix it tonight or tomorrow.

> If you want to add a stacked-area graph to the default dashboard, you'll also want to update this file: https://github.com/DataDog/integrations-core/blob/master/apache/assets/dashboards/apache_dashboard.json

Should I just add it to the end of the existing dashboard with an explanation of what the different states mean (like what already exists with busy and idle workes). Or can I try to create a nicer layout where the new chart won't fall under the fold on a 16:9/16:10 aspect ratio?

Do you think it would help to add a line of max workers to the existing chart with busy and idle workers? Or would it be better to not include that (or put in on a 2nd y axis) because on less used servers the values for busy & idle and max can be quite a bit apart. E.g. 4 busy, 3 idle and 256 or 400 max. (the default values in different modes).

trishankatdatadog reviewed (changes requested):
Agree with @FlorianVeaux: please rebase correctly. Not sure why I got pinged.

FlorianVeaux commented:
@daanleenders What do you think of updating the current busy/idle graph to include all the other types in a stacked-area graph without the open slot ones? This should be more readable this way

ruthnaebeck reviewed (changes requested):
Docs review

daanleenders commented:
> @daanleenders What do you think of updating the current busy/idle graph to include all the other types in a stacked-area graph without the open slot ones? This should be more readable this way

I added a change to the default dashboard that does this. I also updated the explanation of busy and idle that was next to it. 

To me this would be more valuable but it might overwhelm a novice Apache user, where busy and idle (as well as open, and max) workers are perhaps easier to understand? We could rearrange the existing widgets a little bit and use the lowest row for 3 timeseries; the existing one, the more detailed one and busy vs max as a percentage (or something similar)

proposal: 
![image](https://user-images.githubusercontent.com/4190722/119687832-ca3ec980-be47-11eb-8cc5-73699cac28d3.png)

current:
![image](https://user-images.githubusercontent.com/4190722/119687893-d6c32200-be47-11eb-9594-0dd1e0751f86.png)

committed in the PR:
![image](https://user-images.githubusercontent.com/4190722/119688987-c6f80d80-be48-11eb-9aeb-93a4de1bb26c.png)

_I didn't have the new metrics loaded in the account I used to create these examples that's why the new chart is empty._

ruthnaebeck commented on apache/assets/dashboards/apache_dashboard.json:
```diff
@@ -451,7 +497,7 @@
             "definition": {
                 "background_color": "gray",
-                "content": "A worker is considered \u201cbusy\u201d if it is in any of the following states: reading, writing, keep-alive, logging, closing, or gracefully finishing.\n\nAn \u201cidle\u201d worker is not in any of the busy states; the number of idle workers is shown on Apache\u2019s mod_status page.",
+                "content": "A worker is considered \u201cbusy\u201d if it is in any of the following states: Reading request, Sending reply, Keep-alive, Logging, Closing connection, or DNS lookup.\n\nAn worker in the \u2019Waiting for connection\u2019 state is considered \u201cidle\u201d.\n\n; The \u2019Gracefully finishing\u2019 state counts as \u201cidle\u201d when you are in async mode (you se ...
```
```suggestion
                "content": "A worker is considered \u201cbusy\u201d if it is in any of the following states: Reading request, Sending reply, Keep-alive, Logging, Closing connection, or DNS lookup.\n\nA worker in the \u2019Waiting for connection\u2019 state is considered \u201cidle\u201d.\n\n; The \u2019Gracefully finishing\u2019 state counts as \u201cidle\u201d when you are in async mode (you see the async metrics) but as \u201cbusy\u201d in any other mode.\n\n Workers in the \u2019Idle cleanup\u2109 state are moving from an idle state to an open/unused slot. \n\n A \u2109Disabled\u2109 worker can never be used in any of the other states, check your mpm worker config to allow them to be used.",
```

ruthnaebeck commented on apache/metadata.csv:
```diff
@@ -10,4 +10,17 @@ apache.net.request_per_s,gauge,,request,second,The number of requests performed
+apache.scoreboard.waiting_for_connection,gauge,,thread,,The number of workers that can immediately process an incoming request.,0,apache,waiting for connection
+apache.scoreboard.starting_up,gauge,,thread,,The workers that are still starting up and not yet able to handle a request.,0,apache,starting up
+apache.scoreboard.reading_request,gauge,,thread,,The workers reading the incoming request.,0,apache,reading request
+apache.scoreboard.sending_reply,gauge,,thread,,The number of workers sending a reply/response or waiting on a script (like php) to finish so it can send a reply.,0,apache,sending reply
```
```suggestion
apache.scoreboard.sending_reply,gauge,,thread,,The number of workers sending a reply/response or waiting on a script (like PHP) to finish so they can send a reply.,0,apache,sending reply
```

ruthnaebeck commented on apache/metadata.csv:
```diff
@@ -10,4 +10,17 @@ apache.net.request_per_s,gauge,,request,second,The number of requests performed
+apache.scoreboard.keepalive,gauge,,thread,,"The workers intended for a new request from the same client, because it asked to keep the connection alive.",0,apache,keepalive
+apache.scoreboard.dns_lookup,gauge,,thread,,The workers waiting on a DNS lookup.,0,apache,dns lookup
+apache.scoreboard.closing_connection,gauge,,thread,,The amount of workers that are currently closing a connection.,0,apache,closing connection
+apache.scoreboard.logging,gauge,,thread,,The workers writing something to the apache logs.,0,apache,logging
```
```suggestion
apache.scoreboard.logging,gauge,,thread,,The workers writing something to the Apache logs.,0,apache,logging
```

ruthnaebeck commented on apache/metadata.csv:
```diff
@@ -10,4 +10,17 @@ apache.net.request_per_s,gauge,,request,second,The number of requests performed
+apache.scoreboard.logging,gauge,,thread,,The workers writing something to the apache logs.,0,apache,logging
+apache.scoreboard.gracefully_finishing,gauge,,thread,,The number of workers finishing their request.,0,apache,gracefully finishing
+apache.scoreboard.idle_cleanup,gauge,,thread,,These workers were idle and their process is being stopped.,0,apache,idle cleanup
+apache.scoreboard.open_slot,gauge,,thread,,The amount of workers that apache can still start before hitting the maximum number of workers.,0,apache,open slots
```
```suggestion
apache.scoreboard.open_slot,gauge,,thread,,The amount of workers that Apache can still start before hitting the maximum number of workers.,0,apache,open slots
```

ruthnaebeck reviewed (changes requested):
Docs review

daanleenders commented:
> Docs review

This is where English as a second language reveals itself 😄

FlorianVeaux reviewed (approved):
Perfect, thanks @daanleenders !
We're currently on a release freeze until next week or so. I'll merge your PR soon after 👍

daanleenders commented:
> 
> 
> Perfect, thanks @daanleenders !
> We're currently on a release freeze until next week or so. I'll merge your PR soon after 👍

Awesome, I wanted this for months until I finally decided to try and build it myself so I am perfectly happy with waiting another week or so. 

Do you immediately tag/make a new release after merging a PR or do I have to wait until a specific release moment before it is in a agent build I can update to in production?

FlorianVeaux commented:
We can release the integration separately from the agent as soon as the PR is merged, you'll then be able to install it manually with these instructions: https://docs.datadoghq.com/agent/guide/integration-management/
And of course, it will be included in the next agent release

daanleenders commented:
O cool, didn't know you could do that. I'll wait until this can be merged and update the integration on my servers 😄
